[API Compatibility] Add paddle.compat.nn.functional.sdpa

LittleHeroZZZX opened the pull request:
<!-- TemplateReference: https://github.com/PaddlePaddle/Paddle/wiki/PULL-REQUEST-TEMPLATE--REFERENCE -->
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->

### PR Category
<!-- One of [ User Experience | Execute Infrastructure | Operator Mechanism | CINN | Custom Device | Performance Optimization | Distributed Strategy | Parameter Server | Communication Library | Auto Parallel | Inference | Environment Adaptation ] -->
User Experience

### PR Types
<!-- One of [ New features | Bug fixes | Improvements | Performance | BC Breaking | Deprecations | Docs | Devs | Not User Facing | Security | Others ] -->
New features

### Description
<!-- Describe what you’ve done -->
新增 paddle.nn.functional.scaled_dot_product_attention API，涉及改动如下：
- 新增 compat API: compat 与 原版diff 主要在于**输入输出 layout、scale和 gqa**。
  - layout 通过在调用原版 api 前后新增转置处理解决
  - scale 通过让增强原版 API 功能解决
  
- 为原版 sdpa 新增 scale GQA 参数
  - GQA 通过 repeat heads 转换为普通 GQA
  - scale 参数调用增强后的 paddle.compat.F.scaled_dot_production_attention 自动调度

本地实测与 torch 实现可在 1e-2 精度对齐

#76471 中遗留 TOOD：
- [x] weight 和 bias 位置 https://github.com/PaddlePaddle/Paddle/pull/76471#discussion_r2546570885
- [x] compat/nn/init.py 中 \_\_all\_\_ 覆盖 https://github.com/PaddlePaddle/Paddle/pull/76471#discussion_r2548430267
- [x] MHA 中启用 sdpa

LittleHeroZZZX committed:
Implement paddle.nn.functional.sdpa

LittleHeroZZZX committed:
Enable flash attention test and disable test_compat_attention on Windows

LittleHeroZZZX commented:
/re-run all-failed

zhwesky2010 commented on python/paddle/compat/nn/functional/attention.py:
```diff
@@ -0,0 +1,161 @@
+            )
+        elif attn_mask.ndim == 2:
+            # (L, S) -> (1, 1, L, S)
+            attn_mask = attn_mask.unsqueeze(0).unsqueeze(0)
```
x.expand是不是就能解决问题

LittleHeroZZZX replied:
Done

zhwesky2010 commented on python/paddle/nn/functional/flash_attention.py:
```diff
@@ -1455,15 +1485,15 @@ def scaled_dot_product_attention(
+    # flash attention V2 does not support scale
+    if scale is None and attn_mask is None:
         # downgraded to ordinary flash attention implementation
         out, _ = flash_attention(query, key, value, dropout_p, is_causal)
```
flash_attention支持softmax_scale参数。

mem_efficient、math、flash_attn分支都能支持scale。flash_attention里面适配下就行。直接传入scale到flash_attention。

LittleHeroZZZX replied:
flash_attention 里面有个 FLAG 控制 FA 的版本，目前框架用的是FA2，不支持scale参数

zhwesky2010 commented on python/paddle/nn/functional/flash_attention.py:
```diff
@@ -1455,15 +1485,15 @@ def scaled_dot_product_attention(
 
-    if attn_mask is None:
+    # flash attention V2 does not support scale
+    if scale is None and attn_mask is None:
```
可以给flash_attention适配下不同分支下的scale，然后直接传给flash_attention。目前只有flash_attn2不支持。

zhwesky2010 commented on python/paddle/nn/functional/flash_attention.py:
```diff
@@ -1530,14 +1560,19 @@ def scaled_dot_product_attention(
                 [query.shape[1]] * query.shape[0], dtype='int32'
             )
 
```
这里传入的scale要么是默认值1.0 / np.sqrt(query.shape[-1])，要么是用户传入的，这里不需要删掉吧。

建议在最前面直接取默认值
```
if scale is None:
    scale = ...
```

LittleHeroZZZX replied:
默认值下一层调用自己会计算，这里反而会覆盖掉用户指定的值

zhwesky2010 commented on python/paddle/nn/functional/flash_attention.py:
```diff
@@ -1455,15 +1485,15 @@ def scaled_dot_product_attention(
         out, _ = flash_attention(query, key, value, dropout_p, is_causal)
     else:
-        head_dim = query.shape[3]
         sdp_func_name = _select_sdp_for_sdpa(
```
这里是不是不用select，flash_attn分支下直接调用flash_attention来处理scale，其他分支都可支持scale。

zhwesky2010 commented on tools/windows/run_unittests.sh:
```diff
@@ -170,6 +170,7 @@ disable_wingpu_cuda12_test="^test_cholesky_op$|\
 ^test_conv3d_layer$|\
 ^test_decorator$|\
 ^test_flash_attention$|\
+^test_compat_attention$|\
```
这个需要禁用单测吗

LittleHeroZZZX replied:
test_flash_attention 这个单测在 win 上也是禁用的，打开之后报错没有 FA 和 Mem Efficient attention 的实现。不确定是不是 Win 上不支持？

<img width="1914" height="2052" alt="Image" src="https://github.com/user-attachments/assets/095adc66-ee22-4ed3-964b-171eb071f534" />

LittleHeroZZZX replied:
另外 converage 覆盖率也没过，怀疑是 converge cuda 版本太低直接跳过了这个测试？这种情况咋处理？

zhwesky2010 commented:
@LittleHeroZZZX 这个按上次的方式还能改不。

LittleHeroZZZX committed:
Refactor sdpa

LittleHeroZZZX commented:
/re-run all-failed

LittleHeroZZZX commented on test/legacy_test/test_flash_attention.py:
```diff
@@ -1657,6 +1684,109 @@ def setUp(self):
             self.mask[i, 0, :seq_len, :seq_len] = mask * 1e4
         self.rtol = 1e-3
         self.atol = 1e-3
+        self.expected_output_without_mask = np.array(
```
FA 不支持传入自定义 mask，当 is_causal 时，不传入 mask 以选择 FA 后端

LittleHeroZZZX commented on python/paddle/nn/functional/sdpa.py:
```diff
@@ -0,0 +1,721 @@
+        fixed_seed_offset = None
+        return_softmax = False
+        rng_name = ""
+        out, _, _, _ = _C_ops.flash_attn(
```
根据之前的讨论，当选到 FA 后端时，我们直接调 Kernel，而不是 flash_attention 中提供的 flash_attention 接口（它还会重新选择后端，而不是一定走 FA）

LittleHeroZZZX commented on python/paddle/nn/functional/scaled_dot_production_attention.py:
```diff
@@ -0,0 +1,721 @@
+        if not constraint(params, debug):
+            return False
+
+    if not check_scale_is_None(params, debug):
```
这里与torch 存在 diff：我们的 FA 在 kernel 里硬编码了 scale，或许可以通过修改 kernel 对齐

zhwesky2010 commented on python/paddle/compat/nn/functional/attention.py:
```diff
@@ -0,0 +1,171 @@
+                value.flatten(-4, -3).contiguous(),
+            )
+
+    if attn_mask is not None:
```
这里面compat的有些逻辑，如果是兼容的，直接放到原sdpa里一起使用吧。

例如enable_gqa、attn_mask的逻辑可以放原sdpa逻辑里，compat.sdpa只处理两者不可绕过的不兼容差异：输入shape不同

LittleHeroZZZX replied:
mask 的处理已经移除了，sdpa 可以考虑增强一下，支持 enable_gqa

zhwesky2010 commented on python/paddle/nn/functional/flash_attention.py:
```diff
@@ -227,16 +122,39 @@ def _math_attention(
     """
+    q_heads = query.shape[2]
+    k_heads = key.shape[2]
+    if q_heads != k_heads:
```
这个enable_gqa逻辑是不是在最前面应统一处理好。

F.sdpa统一处理enable_gqa、scale这些公共逻辑，compat.sdpa和下游的后端不再处理这些公共逻辑。

LittleHeroZZZX replied:
嗯，这部分逻辑已经冗余了，math 不需要处理 gqa。

我觉得 scale 还是直接让 sdpa 穿透转发比较合适，传 None 表示使用标准 attention 公式里的scale。

zhwesky2010 replied:
> 嗯，这部分逻辑已经冗余了，math 不需要处理 gqa。
> 
> 我觉得 scale 还是直接让 sdpa 穿透转发比较合适，传 None 表示使用标准 attention 公式里的scale。

那这部分expand逻辑还需要吗，看还在？

zhwesky2010 commented on python/paddle/nn/functional/flash_attention.py:
```diff
@@ -246,7 +164,11 @@ def _math_attention(
-        if "xpu" in place:
+        if (
+            "xpu" in place
+            or product.shape[-1] < 32
```
这个新增的条件是为对齐torch吗

LittleHeroZZZX replied:
下面这个融合的分支 softmax_mask_fuse_upper_triangle 算子只支持 32 ~ 16384：
https://github.com/PaddlePaddle/Paddle/blob/8936a3139aa930877dd926c00814738d998be626/paddle/phi/kernels/fusion/gpu/fused_softmax_mask_upper_triangle_kernel.cu#L177-L182

zhwesky2010 commented on python/paddle/nn/functional/sdpa.py:
```diff
@@ -0,0 +1,721 @@
+
+
+def can_use_flash_attention(params: SDPParams, debug: bool = False) -> bool:
+    general_constraints = [
```
这里看条件变多了？这里的条件有额外新增的吗，有精简删去的吗

LittleHeroZZZX replied:
这里对齐了 torch，相比原 sdpa，增加了：
- check_tensor_shapes，确保输入都是四维 Tensor
- check_for_attn_mask，原 sdpa 在 select_sdp 做了前置判断
- check_requires_grad_and_head_dim_gt192_constraints_on_sm86_89_or_120，若需要计算梯度，在特定架构（消费级40系显卡和blackwell）上限制 head_dim 范围 https://github.com/pytorch/pytorch/blob/2f023bf7b962e69c0de01b89af197388d9df27cc/aten/src/ATen/native/transformers/cuda/sdp_utils.cpp#L405-L437
- check_flash_causal_non_square_seqlens 检查 qk 长度一致，FAV2 使用的是正方形 mask
- check_batch_size_and_num_heads_dense 检查 qkv batchsize 一致
- check_nonzero_sequence_lengths_dense 确保 qk 的长度不为 0 
- check_last_dim_stride_equals_1_dense 确保 qkv 在最后一个维度上是连续的

LittleHeroZZZX committed:
check dtype for mem_efficient_attention, support 3d attn_mask, refine tests

zhwesky2010 commented on python/paddle/nn/functional/scaled_dot_production_attention.py:
```diff
@@ -0,0 +1,721 @@
+    if len(config) == 0:
+        init_config()
+
+    if attn_mask is not None and attn_mask.dtype == paddle.bool:
```
attn_mask不同地方都有处理，scale、gqa、mask以及其他的公共处理逻辑都收到sdpa入口这来吧。上游compat、下游后端都复用。

LittleHeroZZZX replied:
Done

LittleHeroZZZX replied:
scale 不需要做预处理，透传即可

zhwesky2010 commented on python/paddle/nn/functional/scaled_dot_production_attention.py:
```diff
@@ -0,0 +1,721 @@
+        param,
+        debug=debug_sdpa,
+    )
+    print("Selected backend", sdp_func_name)
```
调试信息可删除

LittleHeroZZZX replied:
Done，通过 debug flag 控制

zhwesky2010 commented on python/paddle/nn/functional/sdpa.py:
```diff
@@ -0,0 +1,721 @@
+        )
+
+        if is_causal:
+            bias_input = LowerTriangularMask()
```
attn_bias

LittleHeroZZZX replied:
当传入 is_causal = True 时， attn_mask 为 None

zhwesky2010 commented on python/paddle/compat/nn/functional/attention.py:
```diff
@@ -0,0 +1,171 @@
+        raise RuntimeError(
+            "Explicit attn_mask should not be set when is_causal=True"
+        )
+    if len(query.shape) == 3:
```
这些兼容逻辑都可以挪到F.sdpa里，后面有对shape的处理。这里也不用判断两次了。

LittleHeroZZZX replied:
这个改动不兼容，原来的sdpa 如果同时传入 mask 和 is_causal=True，不会抛出异常，还是放在compat 更合适？

zhwesky2010 commented on python/paddle/nn/functional/scaled_dot_production_attention.py:
```diff
@@ -0,0 +1,721 @@
+            bias_input = attn_mask
+        else:
+            bias_input = None
+        if isinstance(bias_input, paddle.Tensor) and bias_input.ndim == 4:
```
attn_bias=attn_mask后，attn_mask在上面公共逻辑处理完后，这里还有额外逻辑吗

LittleHeroZZZX replied:
已移除

LittleHeroZZZX committed:
fix test_flash_attn error

LittleHeroZZZX committed:
feat: refactor GQA implementation and improve tensor handling

- Move GQA logic from compat module to main scaled_dot_product_attention function
- Add enable_gqa parameter to function signature with proper documentation
- Simplify tensor dimension handling using is_batched check
- Remove duplicate GQA validation and expansion code from compat module
- Improve code organization by centralizing GQA functionality in main implementation

LittleHeroZZZX committed:
feat(test): update attention test shape for better alignment

Change the shape parameter in TestSDPAttentionWithScale from (2, 32, 8, 32) to (2, 8, 8, 32) to improve test alignment and ensure proper attention mechanism validation with more realistic tensor dimensions.

zhwesky2010 commented on python/paddle/nn/functional/scaled_dot_production_attention.py:
```diff
@@ -0,0 +1,771 @@
+    return paddle.is_compiled_with_cuda() and paddle.device.is_available()
+
+
+def check_all_tensors_on_device(params: SDPParams, debug: bool):
```
可以根据paddle特性来优化下判断。paddle一般所有tensor在同一个place上，这个判断可以精简掉

LittleHeroZZZX replied:
这个还是有必要的，它不是检查都在一个 place 上（名字有点误导，来自 torch），而是检查输入的设备是否是 GPU，CPU 的 Tensor 还是得分发到 math 的实现上。

FA 精简了 如下检查：
1.  check_tensor_shapes，检查 qkv 是否是四维 Tensor，在 F.sdpa 中会把三维 unsqueeze 到四维，条件恒真
2. 修改 check_flash_attention_hardware_support，直接传入 device id，并缓存结果，从而避免对同一个 device 反复检查
3. check_requires_grad_and_head_dim_gt192_constraints_on_sm86_89_or_120，flashattention 2.5.5 开始在所有设备上均支持最大 256 的反向 https://github.com/Dao-AILab/flash-attention#nvidia-cuda-support
4. check_batch_size_and_num_heads_dens，检查 batchsize 和 num_heads 一致，在 sdpa 中已经预处理过了
5. check_nonzero_sequence_lengths_dense，query 或者 key 的 seq_len 是否为 0，我们的能正确处理，可以跳过
6. check_last_dim_stride_equals_1_dense 检查最后一个维度的步长是否为 0，我们的能自动转连续，跳过

MemEfficientAttn 精简了如下检查：
1. 修改 check_mem_efficient_hardware_support，直接传入 device id，并缓存结果，从而避免对同一个 device 反复检查

zhwesky2010 commented on python/paddle/nn/functional/scaled_dot_production_attention.py:
```diff
@@ -0,0 +1,771 @@
+    return True
+
+
+def check_attn_mask_alignment(params: SDPParams, debug: bool) -> bool:
```
目前就是感觉条件实在太多了。最好根据paddle特性来精简下，依据paddle的flash_attn、memory_efficient实际能力来决定后端。如果flash_attn能计算，还是可以分发到flash_attn后端的。

参考下flash_attention.py看看哪些是必要条件留下来。

LittleHeroZZZX replied:
Done, 具体精简的条件如上

LittleHeroZZZX committed:
feat(attention): update documentation and mask handling logic

- Update scaled_dot_product_attention documentation to clarify dtype support and remove GQA mode mention
- Simplify mask padding logic in MultiheadAttention to always use input dtype
- Add tensor shape comments for better code readability
- Refactor attention mask generation logic to improve efficiency
- Remove unused device capability checking functions

These changes improve code clarity and maintainability while ensuring consistent behavior across different input types.

LittleHeroZZZX committed:
feat(transformer): initialize bias parameters with None and conditionally create bias parameters

Initialize all bias parameters (in_proj_bias, q_proj_bias, k_proj_bias, v_proj_bias) to None at class initialization. Conditionally create bias parameters only when bias=True, moving the bias parameter creation logic to the appropriate conditional branches. This improves code clarity by ensuring bias parameters are always defined and only created when needed.

LittleHeroZZZX committed:
feat(nn): remove  __all__ from compat nn module

LittleHeroZZZX committed:
feat: fix CUDA availability check in scaled dot product attention

Change `paddle.device.is_available()` to `paddle.cuda.is_available()` in the CUDA availability check function. This ensures proper detection of CUDA availability specifically for GPU operations in the scaled dot product attention implementation.

LittleHeroZZZX commented:
/re-run all-failed

zhwesky2010 commented on python/paddle/compat/nn/functional/sdpa.py:
```diff
@@ -0,0 +1,126 @@
+# Copyright (c) 2025 PaddlePaddle Authors. All Rights Reserved.
```
新老sdpa的文件名都统一叫sdpa.py吧

```
新scaled_dot_product_attention：python/paddle/compat/nn/functional/sdpa.py
老scaled_dot_product_attention：python/paddle/nn/functional/sdpa.py
```

zhwesky2010 commented on python/paddle/nn/functional/scaled_dot_production_attention.py:
```diff
@@ -0,0 +1,663 @@
+    dropout_p: float = 0.0,
+    is_causal: bool = False,
+    training: bool = True,
+    name: str | None = None,
```
这个name没用，可以挪最后面

zhwesky2010 commented on python/paddle/nn/functional/scaled_dot_production_attention.py:
```diff
@@ -0,0 +1,663 @@
+        if not constraint(params, debug):
+            return False
+
+    if not check_flash_attention_hardware_support(params.device_id[0], debug):
```
这两个条件：check_flash_attention_hardware_support、check_scale_is_None，为何不统一放上面constraints list

LittleHeroZZZX replied:
check_scale_is_None 是我们独有的条件，就没放进去，这个可以放进去。
check_flash_attention_hardware_support 只和每张卡有关，为了能缓存每张卡的结果，只传了卡的id而不是整个param

zhwesky2010 commented on python/paddle/nn/functional/scaled_dot_production_attention.py:
```diff
@@ -0,0 +1,663 @@
+    for constraint in constraints:
+        if not constraint(params, debug):
+            return False
+    if not check_mem_efficient_hardware_support(params.device_id[0], debug):
```
这个条件check_mem_efficient_hardware_support，为何不统一放上面list

LittleHeroZZZX replied:
同上

zhwesky2010 commented on python/paddle/nn/functional/scaled_dot_production_attention.py:
```diff
@@ -0,0 +1,663 @@
+        attn_mask = attn_mask.expand(mask_shape)
+    sdp_func_name = select_sdp_for_sdpa(param, debug_sdpa)
+
+    if debug_sdpa:
```
debug_sdpa这个用户能传入吗，如果用户没有入口控制，可以考虑logging.info这种

python/paddle/base/backward.py

<img width="629" height="73" alt="infoflow 2025-11-24 16-57-19" src="https://github.com/user-attachments/assets/4d0f775d-9aad-471a-9e01-3b01b35a5430" />

LittleHeroZZZX replied:
Done

zhwesky2010 commented on test/legacy_test/test_compat_scaled_dot_product_attention.py:
```diff
@@ -0,0 +1,331 @@
+# Copyright (c) 2025 PaddlePaddle Authors. All Rights Reserved.
```
命名和老版sdpa对应，就叫`test_compat_scaled_dot_product_attention.py` 吧

LittleHeroZZZX replied:
Done

zhwesky2010 reviewed (commented):
冲突了。

有几个地方顺便也改下

LittleHeroZZZX committed:
feat: update shape output format in docstrings and rename attention module

- Change shape output format from list to paddle.Size in AvgPool1D, AvgPool2D, AvgPool3D, and Unfold docstrings
- Rename attention.py to sdpa.py and update import paths
- Remove debug parameter from check_all_tensors_on_device function
- Replace debug warning with info logging for tensor device placement checks
- Update MultiheadAttention documentation regarding optimized implementation conditions

zhwesky2010 reviewed (dismissed):
LGTM

LittleHeroZZZX committed:
feat: reduce log verbosity in attention validation functions

Changed logger calls from info to debug level in SDPA validation functions to reduce noise in production logs. This maintains the same validation logic but only shows detailed validation messages when debug logging is enabled.

LittleHeroZZZX committed:
feat: add bfloat16 support check for MHA tests on CUDA

Add paddle.device.is_bf16_supported() check to ensure bfloat16 tests
only run on CUDA devices that support bfloat16. This prevents test
failures on CUDA devices without bfloat16 support by falling back to
float32 dtype in those cases.

LittleHeroZZZX committed:
feat: add runtime flags for attention backends and fix bf16 support check

- Add FLAGS_memory_efficient_attention_available and FLAGS_flash_attention_available
  to conditionally enable attention backends at runtime
- Update SDPA backend selection to use runtime flags instead of hardcoded values
- Fix bf16 support detection in multihead attention tests by checking CUDA compute capability
- Remove redundant scale check in flash attention constraints
- Improve test coverage by using consistent bf16 capability checks

LittleHeroZZZX committed:
feat: add global flags for attention kernel availability

Add global boolean flags `memory_efficient_attention_available` and `flash_attention_available` to centralize availability checks for memory efficient and flash attention kernels. Move flag definitions from individual kernel files to flags.cc for better maintainability and to avoid code duplication. The flags automatically set to true when corresponding compilation macros (PADDLE_WITH_MEMORY_EFFICIENT_ATTENTION and PADDLE_WITH_FLASHATTN) are defined, allowing runtime detection of available attention implementations.

LittleHeroZZZX committed:
Fix compile error on windows

zhwesky2010 commented on paddle/phi/core/flags.cc:
```diff
@@ -32,3 +32,25 @@ PHI_DEFINE_EXPORTED_int64(conv_workspace_size_limit,
                           "cuDNN convolution workspace limit in MB unit.");
 #endif
+
+#ifdef PADDLE_WITH_MEMORY_EFFICIENT_ATTENTION
```
新增的几个flag的作用是啥

LittleHeroZZZX replied:
在某些平台和设备上不支持FA或者MEA，通过这些 flag 在Python 端获取实际可用的后端

目前已经发现的case ：
- Windows 上编译时不开启 CUTLASS，没有 MEA 后端
- DCU 上不支持 MEA

zhwesky2010 commented on python/paddle/nn/attention/sdpa.py:
```diff
@@ -17,6 +17,7 @@
 from enum import IntEnum
 from typing import TYPE_CHECKING
 
+import paddle
```
后面如果有PR，可以顺便把这个文件名改下：

python/paddle/nn/attention/sdpa.py -> python/paddle/nn/attention/sdpa_utils.py

目前有其他两个sdpa.py了，存量这个不再适合命名sdpa.py了

LittleHeroZZZX replied:
下个 PR 再改，风险有点大😂，不知道哪个单测就导入了这个文件

LittleHeroZZZX committed:
Fix build error

LittleHeroZZZX commented:
/re-run all-failed

LittleHeroZZZX commented:
/re-run all-failed

zhwesky2010 reviewed (dismissed):
LGTM

LittleHeroZZZX committed:
feat(nn): use safe dict get for attention backend flags

Replace direct dictionary access with get() method to handle missing flags gracefully. This prevents KeyError exceptions when the global flags dictionary doesn't contain the expected flash attention and memory efficient attention availability flags, providing default False values instead.

LittleHeroZZZX commented:
/re-run all-failed

XiaoguangHu01 reviewed (approved):
LGTM
